File: demo/src/main/kotlin/com/example/demo/chatgpt/config/SlackConfig.java
package com.example.demo.chatgpt.config;

import static com.example.demo.chatgpt.service.PapagoService.Language.ENGLISH;
import static com.example.demo.chatgpt.service.PapagoService.Language.KOREAN;

import com.example.demo.chatgpt.service.ChatGptService;
import com.example.demo.chatgpt.service.PapagoDetectLanguageService;
import com.example.demo.chatgpt.service.PapagoService;
import com.slack.api.app_backend.slash_commands.payload.SlashCommandPayload;
import com.slack.api.bolt.App;
import com.slack.api.bolt.AppConfig;
import com.slack.api.bolt.util.SlackRequestParser;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class SlackConfig {

    private final PapagoService papagoService;
    private final PapagoDetectLanguageService detectLanguageService;
    private final ChatGptService chatGptService;

    @Value("${slack.token}")
    private String slackToken;
    @Value("${slack.signingSecret}")
    private String signingSecret;
    @Value("${slack.channelId}")
    private String channelId;

    public SlackConfig(PapagoService papagoService, PapagoDetectLanguageService detectLanguageService, ChatGptService chatGptService) {
        this.papagoService = papagoService;
        this.detectLanguageService = detectLanguageService;
        this.chatGptService = chatGptService;
    }

    @Bean
    public App app() {
        AppConfig appConfig = AppConfig.builder().singleTeamBotToken(slackToken).signingSecret(signingSecret).build();
        App app = new App(appConfig);

        app.command("/gpt", (payload, ctx) -> {
            SlashCommandPayload pl = payload.getPayload();
            String msg = pl.getText();

            if (detectLanguageService.detect(msg) == KOREAN) {
                System.out.println("========================");
                System.out.println("this is translate msg~!!");
                System.out.println(msg);
                msg = papagoService.translate(KOREAN, ENGLISH, msg);
                System.out.println(msg);
                System.out.println("========================");
            }
            System.out.println("========================");
            System.out.println("this is translate answer~!!");
            String answer = chatGptService.send(msg);
            System.out.println(answer);
            String koreanAnswer = papagoService.translate(ENGLISH, KOREAN, answer);
            System.out.println(koreanAnswer);
            System.out.println("========================");
            ctx.client().chatPostMessage(r -> r.channel(channelId)
                    .text("<@" + pl.getUserId() + "> : " + pl.getText() + "\r\n\r\n" +
                            koreanAnswer));

            return ctx.ack();
        });

        return app;
    }

    @Bean
    public SlackRequestParser slackRequestParser() {
        return new SlackRequestParser(app().config());
    }
}

File: demo/src/main/kotlin/com/example/demo/chatgpt/config/SlackConfig.kt
package com.example.demo.chatgpt.config

import com.example.demo.chatgpt.service.ChatGptService
import com.example.demo.chatgpt.service.PapagoDetectLanguageService
import com.example.demo.chatgpt.service.PapagoService
import com.slack.api.bolt.App
import com.slack.api.bolt.AppConfig
import com.slack.api.bolt.context.builtin.SlashCommandContext
import com.slack.api.bolt.request.builtin.SlashCommandRequest
import com.slack.api.bolt.util.SlackRequestParser
import com.slack.api.methods.request.chat.ChatPostMessageRequest.ChatPostMessageRequestBuilder
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class SlackConfig(
    private val papagoService: PapagoService,
    private val detectLanguageService: PapagoDetectLanguageService,
    private val chatGptService: ChatGptService
) {
    @Value("\${slack.token}")
    private val slackToken: String? = null

    @Value("\${slack.signingSecret}")
    private val signingSecret: String? = null

    @Value("\${slack.channelId}")
    private val channelId: String? = null

    @Bean
    fun app(): App {
        val appConfig = AppConfig.builder().singleTeamBotToken(slackToken).signingSecret(signingSecret).build()
        val app = App(appConfig)
        app.command("/gpt") { payload: SlashCommandRequest, ctx: SlashCommandContext ->
            val pl = payload.payload
            var msg = pl.text
            if (detectLanguageService.detect(msg) == PapagoService.Language.KOREAN) {
                println("========================")
                println("this is translate msg~!!")
                println(msg)
                msg = papagoService.translate(PapagoService.Language.KOREAN, PapagoService.Language.ENGLISH, msg)
                println(msg)
                println("========================")
            }
            println("========================")
            println("this is translate answer~!!")
            val answer = chatGptService.send(msg)
            println(answer)
            val koreanAnswer = papagoService.translate(PapagoService.Language.ENGLISH, PapagoService.Language.KOREAN, answer)
            println(koreanAnswer)
            println("========================")
            ctx.client().chatPostMessage { r: ChatPostMessageRequestBuilder ->
                r.channel(channelId)
                    .text(
                        """
    <@${pl.userId}> : ${pl.text}
    
    $koreanAnswer
    """.trimIndent()
                    )
            }
            ctx.ack()
        }
        return app
    }

    @Bean
    fun slackRequestParser(): SlackRequestParser {
        return SlackRequestParser(app().config())
    }
}
